test/sample_server.go

```diff
@@ -42,15 +60,22 @@ func (c *ControllerImpl) GetBookByID2(ctx context.Context, id string) (*models.B
 	if err != nil {
 		return nil, err
 	}
-	book, ok := c.books[int64(i)]
-	if !ok {
+	if i >= len(c.books) {
+		return nil, models.Error{}
+	}
+	book := c.books[i]
+	if book == nil {
 		return nil, models.Error{}
 	}
 	return book, nil
 }
 
 // CreateBook creates a book.
 func (c *ControllerImpl) CreateBook(ctx context.Context, input *models.Book) (*models.Book, error) {
+	if int(input.ID) >= len(c.books) {
+		extension := make([]*models.Book, int(input.ID)-len(c.books)+1)
+		c.books = append(c.books, extension...)
```

Comment:
I forced them into an array so that I could get consistent ordering on output, but maybe I could just keep track of the max ID and loop through all possible IDs on output. Which do you think is better?